### PR TITLE
Make CoCoA use cvc5's libgmp path

### PR DESCRIPTION
**Pull request by sorawee**
(Edited)

Prior this PR, CoCoA tries to discover libgmp on its own.
This will usually fail on systems like Mac M1/2, where libgmp via
Homebrew installation is usually located at a non-standard location.
This PR passes an explicit libgmp path that cvc5 already computed
to CoCoA, which fixes the issue.

We need to also manually patch CoCoA because CoCoA is buggy:
it made an assumption that libgmp is statically-linked,
so when cvc5 is compiled in the dynamically linked mode
(i.e. without `--static`), CoCoA will fail to compile.
The patch fixes this issue.

### CODE REVIEW

**Comment by alex-ozdemir** [comment]
Thank you for this PR, Sorawee. It would be great to have the build work properly for new macOS machines. The patch looks good. Once CI passes we'll merge it.

This last CI failure was spurious (#10127, to be merged soon-ish), so I'm re-running CI.

**Review by alex-ozdemir — APPROVED** [review]
:+1:

**Comment by sorawee** [comment]
Thanks! The job to build on MacOS is now passed, but other jobs were not re-run. Would you mind re-running them as well?

**Comment by mpreiner** [comment]
This doesn't look like the correct fix. cvc5 already has a dependency on GMP, so CoCoA should use the GMP version found by the cvc5 build system. This means that the cvc5 build system should pass communicate the found GMP version to the CoCoA build system.

**Comment by alex-ozdemir** [comment]
> cvc5 already has a dependency on GMP, so CoCoA should use the GMP version found by the cvc5 build system.

Good point, Mathias.

@sorawee : CoCoA's configuration script has a `--with-gmplib=PATH` argument. You should be able to set it here: https://github.com/cvc5/cvc5/blob/main/cmake/FindCoCoA.cmake#L64.

For an example of a different dependency (not CoCoA) that uses the cvc5-determined GMP library path, see this file: https://github.com/cvc5/cvc5/blob/main/cmake/FindPoly.cmake

**Comment by sorawee** [comment]
That kind of works, and I agree it's a much better method. Thanks!

I said "kind of" because it only works in conjunction with the `--static` flag. When `--static` is provided, `${GMP_LIBRARIES}` is set to `/opt/homebrew/lib/libgmp.a`, which is also the value that my previous method will use. 

However, if `--static` is not provided, `${GMP_LIBRARIES}` is set to `/opt/homebrew/lib/libgmp.dylib`. This then causes the following error:

```
   './configure' '--prefix=/Users/sorawee/projects/solvers/cvc5/build/deps' '--with-libgmp=/opt/homebrew/lib/libgmp.dylib'

  See also

    /Users/sorawee/projects/solvers/cvc5/build/deps/src/CoCoA-EP-stamp/CoCoA-EP-configure.log


-- Log output is:
Starting configuration process for CoCoALib...

./configure: ERROR:  Supplied value of CXXFLAGS, namely "  -std=c++14  -Wall  -pedantic  -fPIC  -O2"
./configure: ERROR:  is not compatible with GMP library
ERROR: failed to compile/link test-gmp-cxxflags --> see LogFile.   [[gmp-check-cxxflags.sh]]
```

**Comment by sorawee** [comment]
On further investigation, CoCoA symlinked `/opt/homebrew/lib/libgmp.dylib` (resp. `/opt/homebrew/lib/libgmp.a`) to `build/deps/src/CoCoA-EP/configuration/ExternalLibs/lib/libgmp-symlink.dylib` (resp. `build/deps/src/CoCoA-EP/configuration/ExternalLibs/lib/libgmp-symlink.a`). 

The compilation flag is then computed from:

```
GMP_LDLIB=-lgmp
if [ -f "$COCOA_EXTLIB_DIR"/lib/libgmp-symlink.a ]
then
  GMP_LDLIB=-lgmp-symlink
fi
```

But this is incorrect. `GMP_LDLIB` should also be set to `-lgmp-symlink` if `build/deps/src/CoCoA-EP/configuration/ExternalLibs/lib/libgmp-symlink.dylib` is present.

So I think the fix is to patch `CoCoA-EP/configuration/gmp-check-cxxflags.sh` to also look for `"$COCOA_EXTLIB_DIR"/lib/libgmp-symlink.(so|dylib)`.

**Comment by sorawee** [comment]
I pushed a fix. Could anyone run GHA, please? Thank you!

**Comment by sorawee** [comment]
I just found that the above issue was already reported upstream at https://cocoa.dima.unige.it/redmine/issues/1749, but there's no activity on the issue so far. The report was also focused on CoCoALib-0.99850, which is a newer version, and has even more issues compared to the version that we are currently using.

**Comment by alex-ozdemir** [comment]
@mpreiner : what do you think of this approach?

**Comment by mpreiner** [comment]
LGTM, thanks for the changes!

**Comment by alex-ozdemir** [comment]
Paging @mpreiner or @ajreynol  to merge this (unless we're waiting on something).

**Comment by mpreiner** [comment]
@sorawee Please sign your commit as explained in https://github.com/cvc5/cvc5/blob/main/CONTRIBUTING.md.

**Comment by sorawee** [comment]
Sorry! I now signed the commit.

**Pull request merged**
